### PR TITLE
Add new parameters to compile-tool

### PR DESCRIPTION
**Pull request by mikhailk62**
Add new parameters to compile-tool

### CODE REVIEW

File: inference-engine/tools/compile_tool/main.cpp [hunk 1]
```diff
@@ -216,6 +238,16 @@ static InferenceEngine::Precision getOutputPrecision(const std::string &value) {
     return getPrecision(value, supported_precisions, "for output layer");
 }
 
+static InferenceEngine::Layout getLayout(const std::string &value) {
+    static const supported_layouts_t supported_layouts = {
+            { "NCHW", InferenceEngine::Layout::NCHW },
+            { "NHWC", InferenceEngine::Layout::NHWC },
+            { "NC", InferenceEngine::Layout::NC },
+            { "c", InferenceEngine::Layout::C }
```
Comment (vinograd47) [inline inference-engine/tools/compile_tool/main.cpp:246]: Upper-case `C`.

Comment (mikhailk62) [reply inference-engine/tools/compile_tool/main.cpp]: Fixed.

File: inference-engine/tools/compile_tool/main.cpp [hunk 2]
```diff
@@ -310,6 +342,23 @@ static void processPrecisions(InferenceEngine::CNNNetwork &network,
     }
 }
 
+static void processLayout(InferenceEngine::CNNNetwork &network,
+                          const std::string &inputs_layout, const std::string &outputs_layout) {
+    if (!inputs_layout.empty()) {
+        auto layout = getLayout(inputs_layout);
+        for (auto &&layer : network.getInputsInfo()) {
+            layer.second->setLayout(layout);
```
Comment (vinograd47) [inline inference-engine/tools/compile_tool/main.cpp:350]: Please check that layout is applicable, eg. it corresponds to the number of dimensions.

Comment (mikhailk62) [reply inference-engine/tools/compile_tool/main.cpp:350]: This check already exists inside the `setLayout` function.

Comment (vinograd47) [reply inference-engine/tools/compile_tool/main.cpp:350]: No, `setLayout` just checks that `std::string` value can be converted into Layout enumeration. I was talking about layout applicability to tensor. For example, you can't set NC layout for 4D tensor.

Comment (vinograd47) [reply inference-engine/tools/compile_tool/main.cpp:350]: You need to get number of dimensions from `layout` and call `layer.second->setLayout` only if the number of dimensions matches the `layer.second->getDims().size()`.

File: inference-engine/tools/compile_tool/main.cpp [hunk 3]
```diff
@@ -289,7 +321,7 @@ static void processPrecisions(InferenceEngine::CNNNetwork &network,
         for (auto &&layer : network.getInputsInfo()) {
             const auto layerPrecision = layer.second->getPrecision();
             if ((isFloat(layerPrecision) && isFloat(precision)) ||
-                (isFP16(layerPrecision) && isU8(precision))) {
+                (isFloat(layerPrecision) && isU8(precision))) {
```
Comment (ArtemySkrebkov-intel) [inline inference-engine/tools/compile_tool/main.cpp:324]: Just wondering why did you add this change?

Comment (vinograd47) [reply inference-engine/tools/compile_tool/main.cpp:324]: Without it you can't set U8 precision for FP32, since by default input will always have FP32 precision (even for FP16 networks).

Comment (vinograd47) [reply inference-engine/tools/compile_tool/main.cpp:324]: Original goal of this check was, I believe, that we can set U8 or FP16 precision only for FP32 tensor. If the tensor is INTEGER, we can't do this.

File: inference-engine/tools/compile_tool/main.cpp (hunk 2 above)
Comment (mikhailk62) [reply inference-engine/tools/compile_tool/main.cpp:350]: https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/inference_engine/ie_layouts.cpp#L97

Comment (vinograd47) [reply inference-engine/tools/compile_tool/main.cpp:350]: It throws exception. I suggest just not to call `setLayout` in unsupported case (as it is done for precision). For example, we have networks with 2 inputs 4D and 1D. Passing `-il NHWC` means that we should change layout only for the first input and do nothing with the second one.

Comment (mikhailk62) [reply inference-engine/tools/compile_tool/main.cpp]: Done.

File: inference-engine/tools/compile_tool/main.cpp (hunk 3 above)
Comment (ArtemySkrebkov-intel) [reply inference-engine/tools/compile_tool/main.cpp:340]: Got it. Thanks!

File: inference-engine/tools/compile_tool/main.cpp [hunk 4]
```diff
@@ -184,6 +193,20 @@ static std::map<std::string, std::string> parsePrecisions(const std::string &iop
 }
 
 using supported_precisions_t = std::unordered_map<std::string, InferenceEngine::Precision>;
+using supported_layouts_t = std::unordered_map<std::string, InferenceEngine::Layout>;
+using matchLayoutToDims_t = std::unordered_map<InferenceEngine::Layout, size_t>;
```
Comment (vinograd47) [inline inference-engine/tools/compile_tool/main.cpp:197]: It looks like maxOS STL doesn't like enums as `std::unordered_map` keys. Please add explicit `std::hash` specialization for `InferenceEngine::Layout` to fix MacOS build.

Comment (mikhailk62) [reply inference-engine/tools/compile_tool/main.cpp]: Fixed.

File: inference-engine/tools/compile_tool/main.cpp [hunk 5]
```diff
@@ -216,6 +239,31 @@ static InferenceEngine::Precision getOutputPrecision(const std::string &value) {
...
+    };
+    return getLayout(value, supported_layouts);
+}
+
+static bool isMatchLayoutToDims(const InferenceEngine::Layout& layout, const size_t dimension) {
+    static const matchLayoutToDims_t matchLayoutToDims = {
+            { std::hash<InferenceEngine::Layout>{}(InferenceEngine::Layout::NCHW), 4 },
+            { std::hash<InferenceEngine::Layout>{}(InferenceEngine::Layout::NHWC), 4 },
+            { std::hash<InferenceEngine::Layout>{}(InferenceEngine::Layout::NC), 2 },
+            { std::hash<InferenceEngine::Layout>{}(InferenceEngine::Layout::C), 1 }};
```
Comment (vinograd47) [inline inference-engine/tools/compile_tool/main.cpp:257]: `static_cast<size_t>` instead of `std::hash<InferenceEngine::Layout>{}`

File: inference-engine/tools/compile_tool/main.cpp [hunk 6]
```diff
@@ -216,6 +239,31 @@ static InferenceEngine::Precision getOutputPrecision(const std::string &value) {
...
+
+    auto dims = matchLayoutToDims.find(layout);
```
Comment (vinograd47) [inline inference-engine/tools/compile_tool/main.cpp:259]: `static_cast<size_t>(layout)`

File: inference-engine/tools/compile_tool/main.cpp (hunk 5 above)
Comment (mikhailk62) [reply inference-engine/tools/compile_tool/main.cpp]: Fixed.

File: inference-engine/tools/compile_tool/main.cpp (hunk 6 above)
Comment (mikhailk62) [reply inference-engine/tools/compile_tool/main.cpp]: Fixed.

**Comment by vinograd47** [comment]
@ilya-lavrenov Please take a look.

File: inference-engine/tools/compile_tool/main.cpp [hunk 7]
```diff
@@ -216,6 +239,31 @@ static InferenceEngine::Precision getOutputPrecision(const std::string &value) {
...
+    static const supported_layouts_t supported_layouts = {
+            { "NCHW", InferenceEngine::Layout::NCHW },
+            { "NHWC", InferenceEngine::Layout::NHWC },
+            { "NC", InferenceEngine::Layout::NC },
+            { "C", InferenceEngine::Layout::C }
+    };
+    return getLayout(value, supported_layouts);
+}
+
+static bool isMatchLayoutToDims(const InferenceEngine::Layout& layout, const size_t dimension) {
```
Comment (briangrifiin) [inline inference-engine/tools/compile_tool/main.cpp:252]: I'd like to propose another name like doesLayoutMatchDims or vice versa (doDimsMatchLayout). That's not a crucial change though.

File: inference-engine/tools/compile_tool/main.cpp [hunk 8]
```diff
@@ -216,6 +239,31 @@ static InferenceEngine::Precision getOutputPrecision(const std::string &value) {
...
+    static const matchLayoutToDims_t matchLayoutToDims = {
+            {static_cast<size_t>(InferenceEngine::Layout::NCHW), 4 },
+            {static_cast<size_t>(InferenceEngine::Layout::NHWC), 4 },
+            {static_cast<size_t>(InferenceEngine::Layout::NC), 2 },
+            {static_cast<size_t>(InferenceEngine::Layout::C), 1 }};
```
Comment (briangrifiin) [inline inference-engine/tools/compile_tool/main.cpp:257]: Do we care about CHW?

Comment (vinograd47) [reply inference-engine/tools/compile_tool/main.cpp:257]: If yes, I'd propose adding 5D layouts as well.

**Comment by vinograd47** [comment]
Please rebase to fix CI build.

**Comment by mikhailk62** [comment]
@ilya-lavrenov ping

**Comment by ilya-lavrenov** [comment]
@mikhailk62 Jenkins CI should pass

File: inference-engine/tools/compile_tool/main.cpp (hunk 8 above)
Comment (briangrifiin) [reply inference-engine/tools/compile_tool/main.cpp:257]: FYI, it seems like one of target networks wants 3D input.

Comment (mikhailk62) [reply inference-engine/tools/compile_tool/main.cpp:259]: Added 3D layout.

**Comment by vinograd47** [comment]
@ilya-lavrenov The CI passed.
